<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title></title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My Website</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">




</div>


<div id="roseq---a-rank-based-approach-to-modeling-gene-expression-in-single-cells" class="section level1">
<h1>ROSeq - A rank based approach to modeling gene expression in single cells</h1>
<p>ROSeq evaluates differential expression between two subpopulations of single cells by modeling the rank distribution of read counts for each gene.</p>
<div id="abstract" class="section level2">
<h2>Abstract</h2>
<p>Single cell transcriptomics provides a window into cell-to-cell variability in complex tissues. Modeling single cell expression is challenging due to high noise levels and technical bias. In the past few years, considerable efforts have been made to devise suitable parametric models for single cell expression data. We use Discrete Generalized Beta Distribution (DGBD) to model the probability distribution of ranks, corresponding to the read counts for a gene and subpopulation. Use of DGBD yields better overall fit across genes compared to the widely used mixture model comprising Poisson and Negative Binomial density functions. Further, we use Wald’s test to probe into differential expression across cell sub-types. We package our implementation as a <strong>standalone</strong> software called ROSeq. When applied on real data-sets, ROSeq performed competitively compared to the state of the art methods including MAST, SCDE and ROTS.</p>
</div>
<div id="downloads" class="section level2">
<h2>Downloads</h2>
<p>The Research Article and Supplementary Material is available for download <a href="https://www.biorxiv.org/content/early/2018/07/22/374025">here</a>.</p>
<p>The R code used for producing differential analysis with methods like SCDE, ROTS, MAST and Wilcoxon-Rank Sum is also available for download in the directory “04_R_code.zip”, which is found in the <strong>Release</strong> section of the repository, in case one wishes to reproduce the results.</p>
</div>
<div id="getting-started" class="section level2">
<h2>Getting Started</h2>
<p>As a first step, download the complete ROSeq <a href="https://github.com/malaalam/ROSeq">repository</a> on to your PC. This would provide the files needed to run the Tutorial Below.</p>
<p>Also, go to the <strong>Release</strong> Section of the repository and download the executables specific to your operating system. Support is provided for Windows, macOS and Linux. Once the zip folder has been downloaded, extract it to a desirable location. Next, open the <strong>for_redistribution</strong> directory and double click the executable file within to begin the installation process.</p>
</div>
<div id="tutorial" class="section level2">
<h2>Tutorial</h2>
<p>This tutorial below, provides instructions to upload the (un) normalized read count data corresponding to single cells coming from the individuals <em>NA19098</em> and <em>NA19101</em>, in the <a href="https://www.ncbi.nlm.nih.gov/pubmed/28045081">tung2017</a> data set; and subsequently analyse the two subpopulations for differential expression.</p>
<p>The number of single cells equals 288 for each individual. After filtering with a threshold equal to six, low quality genes are eliminated and the number of genes considered for differential expression analysis reduces from 19027 to 16087. This is also prompted as a message in the Status Panel.</p>
<p>Following would constitute as a complete set of steps from uploading read count data to saving results as a ’*.csv’ file:</p>
<ol style="list-style-type: decimal">
<li><p>Click on the radio button - <em>Upload Raw Read Count file</em>. This would produce an <em>Open File</em> button, for the user to upload the respective file.</p></li>
<li><p>Click on the <em>Open File</em> button, to select the raw read count file <em>single_counts_no_filter_no_norm.csv</em>. Prior to successful selection of the file, an import wizard window would appear. Click on the buttons <em>Next</em> and <em>Finish</em>.</p></li>
<li><p>Next, the number of genes in the uploaded file will be suggested (19027). The user could alter the number of genes, in case you wish to look only at the first few (10) genes.</p></li>
<li><p>Decide on the Filter Threshold. A filter threshold equal to 6, would mean that all genes with the number of non-zero read counts less than 6, would not be considered during the subsequent differential analaysis, as they would be considered <em>low-quality genes</em>.</p></li>
<li><p>Select the starting (1) and ending (288) column index corresponding to the first group (<em>NA19098</em>). Select the starting (289) and ending (576) column of the second group similarly (<em>NA19101</em>).</p></li>
<li><p>If all steps were performed so far, then the <em>Solve DE</em> Button becomes enabled. Click on it to run the analysis for differential expression. This would open up a green progress bar for the Subpopulation One, followed by another for Subpopulation Two.</p></li>
<li><p>Once the analysis is complete, a message (<em>DE analysis completed successfully.</em>) would appear in the <em>Status</em> Panel. At this stage, the ROSeq software would look as in the Figure below.</p></li>
<li><p>Lastly, click on <em>Save Results</em> button to save the variables [pValue, adjusted pValue and log2(fold count)] as three columns in a *.csv file format.</p></li>
</ol>
<p>[Screenshot of ROSeq]({{ “/05_images/ROSeq.PNG” | absolute_url }})</p>
</div>
<div id="datasets" class="section level2">
<h2>Datasets</h2>
<p>The <a href="https://www.ncbi.nlm.nih.gov/pubmed/28045081">tung2017</a> dataset was downloaded from <a href="https://github.com/jdblischak/singleCellSeq">jdblischak</a>’s Repository. Bulk RNA-seq read count data was extracted from <strong>bulk_reads_per_sample.txt</strong>, while single cell RNA-seq read count data was extracted from <strong>molecules.txt</strong>. (See R code in the Release Subsection of the repository.)</p>
</div>
<div id="troubleshooting" class="section level2">
<h2>Troubleshooting</h2>
<p>In case of any issues or to dive deeper into the implementation, please access the original MATLAB code which has been used to build the Graphical Interface. The file <em>initializeGUI.m</em>, available in the directory <em>03_matlab_files</em> is the entry point to the code. Run this file (press F5 button) and debug! If the problem is not solved through this approach, please feel free to start an <em>issue</em> <a href="https://github.com/malaalam/ROSeq/issues">here</a>.</p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
